### DISCUSSION TITLE
State storage/locality

### DISCUSSION THREAD

**Discussion by orthecreedence** [freenet/freenet-core#1028] [Q&A] [answered]
Forgive me if this is covered in the docs and I missed it, but I'm wondering where the state for a contract is stored? Is it simply stored on the nodes that are operating that particular contract? Or is it stored in some kind of DHT?

I've been evaluating projects like Freenet and [Holochain](https://www.holochain.org/) for building a system where multiple agents can interact with a decentralized system that has "rules" on its own state. For a contrived example, take currency. Let's say a contract represents a shared pool of money. The users authorized to interact with that contract can deposit and withdraw, however if the contract's state-value goes to zero then it would not allow any more withdrawls.

Something like this is difficult in Holochain because all state is stored in the DHT, and each transaction would have to read the DHT (possibly multiple locations) leading to situations where some peer-validators might validate the transaction and others reject it, depending on their view of the DHT at a given time. In other words, to my understanding transactions must be self-contained in Holochain.

So my question extends from a scenario like this...does any given contract have local access to its own state?

**Comment by sanity** [comment]
Contracts and their state are stored in a type of DHT called a small-world network. Contracts control what state is permitted to be associated with them, and the circumstances under which it can be modified.

This is explained in [this talk](https://youtu.be/yBtyNIqZios?si=cMp_zEM8xSzSwJzj) if you have time to watch it, and I just gave a [new talk](https://youtu.be/enTAromEeHo?si=OhrNhBA5_8OwH9RC) in which I go into more detail about how the small-world network works.

I hope that is helpful.

**Reply by orthecreedence** [reply]
I guess what I'm asking is if the contract state always lives locally on any node that holds the contract. If an update to the state happens, does the node the update occurred on modify its state *locally* and then push out a delta to the other holders of the contract? Or does the node make some call out to a separate location on the small-world network to modify the state? Is storage of state separate from the node processing/hosting the contract?

**Reply by orthecreedence** [reply, marked as answer]
Oh, lol, you answered that here: https://youtu.be/enTAromEeHo?si=_-68CiMhz-SIrNz0&t=2917. I had the video running in the background as I typed the above follow up and you answered it not even a minute later. :smile: 

So from my understanding, each peer holding a contract keeps the state locally, and will sync changes to other peers with the same contract.
